numpyro/contrib/tfp/distributions.py

```diff
@@ -187,41 +237,15 @@ def is_discrete(self):
         # XXX: this should cover most cases
         return self.support is None
 
+    def tree_flatten(self):
+        return jax.tree_util.tree_flatten(self.tfp_dist)
 
-InverseGamma = TFPDistribution[tfd.InverseGamma]
-InverseGamma.arg_constraints = {
-    "concentration": constraints.positive,
-    "scale": constraints.positive,
-}
-
-
-def _onehot_enumerate_support(self, expand=True):
-    n = self.event_shape[-1]
-    values = jnp.identity(n, dtype=jnp.result_type(self.dtype))
-    values = values.reshape((n,) + (1,) * len(self.batch_shape) + (n,))
-    if expand:
-        values = jnp.broadcast_to(values, (n,) + self.batch_shape + (n,))
-    return values
-
-
-OneHotCategorical = TFPDistribution[tfd.OneHotCategorical]
-OneHotCategorical.arg_constraints = {"logits": constraints.real_vector}
-OneHotCategorical.has_enumerate_support = True
-OneHotCategorical.support = constraints.simplex
-OneHotCategorical.is_discrete = True
-OneHotCategorical.enumerate_support = _onehot_enumerate_support
-
-OrderedLogistic = TFPDistribution[tfd.OrderedLogistic]
-OrderedLogistic.arg_constraints = {
-    "cutpoints": constraints.ordered_vector,
-    "loc": constraints.real,
-}
-
-Pareto = TFPDistribution[tfd.Pareto]
-Pareto.arg_constraints = {
-    "concentration": constraints.positive,
-    "scale": constraints.positive,
-}
+    @classmethod
+    def tree_unflatten(cls, aux_data, params):
+        fn = jax.tree_util.tree_unflatten(aux_data, params)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=FutureWarning)
+            return cls[fn.__class__](**fn.parameters)
```

Comment:
nit: to trigger the cache, we need to use `TFPDistribution` instead of `cls` (otherwise, a new cache will be stored).